### PR TITLE
Add dweller controller skeleton

### PR DESCRIPTION
**Pull request by kchugalinskiy**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,11 @@
 
 # Project-local glide cache, RE: https://github.com/Masterminds/glide/issues/736
 .glide/
+vendor/
+bin/
+
+# IDEA gitignore
```
Comment (hypnoglow) [inline .gitignore:18]: Use `core.excludesFile` for excluding IDE junk:
https://git-scm.com/docs/gitignore

File: .idea/vcs.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (hypnoglow) [inline .idea/vcs.xml:1]: wtf man

File: README.md [hunk 3]
```diff
@@ -1,2 +1,19 @@
...
+
+```
+go get k8s.io/gengo/examples/deepcopy-gen
+go get k8s.io/apimachinery
+``
+
+Besides, deepcopy-gen requires these packages to exist:
+
+```
+go get golang_org/x/crypto
```
Comment (hypnoglow) [inline README.md:14]: unrecognized import path

File: README.md [hunk 4]
```diff
@@ -1,2 +1,19 @@
...
+go get golang_org/x/net
+```
+
+And, finally, run magic:
+`deepcopy-gen --logtostderr --v=4 -i $(go list ./pkg/apis/dweller/v1alpha1 | paste -sd' ' - | sed 's/ /,/g') -O zz_generated`
```
Comment (hypnoglow) [inline README.md:19]: bad markdown formatting

File: pkg/apis/dweller/controller.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+const maxRetries = 5
+
+type Controller struct {
+	logger    *logrus.Entry
+	clientset kubernetes.Interface
+	queue     workqueue.RateLimitingInterface
+	informer  cache.SharedIndexInformer
+}
+
+func Start() {
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:34]: I think this function contents should be moved to `main` func

File: pkg/apis/dweller/controller.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	config, err := rest.InClusterConfig()
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:35]: Please be friendly to local development:
```go
configPath := flag.String("kubeconfig", "", "Kubernetes config")
flag.Parse()
```

and something like

```go
func GetClientConfig(kubeconfig string) (*rest.Config, error) {
	if kubeconfig != "" {
		return clientcmd.BuildConfigFromFlags("", kubeconfig)
	}
	return rest.InClusterConfig()
}
```

File: pkg/apis/dweller/controller.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		return
+	}
+
+	c := newController(config, kubeClient)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	go c.Run(stopCh)
+
+	sigterm := make(chan os.Signal, 1)
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:53]: this variable should be named like `signals` or `sigs`

File: pkg/apis/dweller/controller.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	signal.Notify(sigterm, syscall.SIGTERM)
+	signal.Notify(sigterm, syscall.SIGINT)
+	<-sigterm
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:56]: This probably should close `stopCh` to gracefully shutdown controller.

File: pkg/apis/dweller/controller.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+	config.APIPath = "/apis"
+	config.GroupVersion = &gv
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: dweller_v1a1.CodecFactory}
+	c, err := rest.RESTClientFor(&config)
+	if nil != err {
+		fmt.Printf("error creating client: %v", err)
+		return nil
+	}
+
+	informer := cache.NewSharedIndexInformer(
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:73]: do we really need a *shared* informer? We are the only controller that uses `secretclaims`

File: pkg/apis/dweller/controller.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+	})
+
+	return &Controller{
+		logger:    logrus.WithField("pkg", "dweller"),
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:122]: hidden dependency :roll_eyes:

File: pkg/apis/dweller/controller.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	go c.Run(stopCh)
+
+	sigterm := make(chan os.Signal, 1)
+	signal.Notify(sigterm, syscall.SIGTERM)
+	signal.Notify(sigterm, syscall.SIGINT)
+	<-sigterm
+}
+
+func newController(k8sConfig *rest.Config, client kubernetes.Interface) *Controller {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:60]: queue should be closer to where it is used

File: pkg/apis/dweller/controller.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+	config := *k8sConfig
+	gv := dweller_v1a1.SchemeGroupVersion
+	config.APIPath = "/apis"
+	config.GroupVersion = &gv
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: dweller_v1a1.CodecFactory}
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:66]: probably a good idea to move config creation logic to a separate function

File: pkg/apis/dweller/controller.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+// Run starts the kubewatch controller
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("Starting dweller controller")
+
+	go c.informer.Run(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, c.HasSynced) {
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:138]: `c.informer.HasSynced`? Why do we need to define `func (c *Controller) HasSynced() bool` ?

File: pkg/apis/dweller/controller.go [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+// HasSynced is required for the cache.Controller interface.
+func (c *Controller) HasSynced() bool {
+	return c.informer.HasSynced()
+}
+
+// LastSyncResourceVersion is required for the cache.Controller interface.
+func (c *Controller) LastSyncResourceVersion() string {
+	return c.informer.LastSyncResourceVersion()
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:155]: please explain why is this needed

File: pkg/apis/dweller/v1alpha1/types.go [hunk 15]
```diff
@@ -0,0 +1,28 @@
+//go:generate codecgen -o types_codec.go types.go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type SecretClaim struct {
```
Comment (hypnoglow) [inline pkg/apis/dweller/v1alpha1/types.go:10]: What do you think if we call this `VaultSecretClaim` (and everywhere else)? Because this controller is designed only for Vault integration.

File: pkg/apis/dweller/controller.go [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+				watcher, err := c.Get().
+					Namespace(meta_v1.NamespaceAll).
+					Resource("secretclaims").
+					VersionedParams(&options, dweller_v1a1.ParameterCodec).
+					Watch()
+				return watcher, err
+			},
+		},
+		&dweller_v1a1.SecretClaim{},
+		0, //Skip resync
```
Comment (hypnoglow) [inline pkg/apis/dweller/controller.go:96]: Why we skip resync? Just for now or we will never need a resync?

File: pkg/apis/dweller/v1alpha1/types.go [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+type SecretClaimStatus struct {
+	Satisfied bool `json:"satisfied"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type SecretClaimList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
```
Comment (hypnoglow) [inline pkg/apis/dweller/v1alpha1/types.go:25]: also `omitempty`

**Review by hypnoglow — CHANGES_REQUESTED** [review]

File: pkg/apis/dweller/controller.go (hunk 12 above)
Comment (kchugalinskiy) [reply pkg/apis/dweller/controller.go:66]: ho-ho, prepare yourself for canonical kubernetes api client ;-)

File: pkg/apis/dweller/controller.go (hunk 9 above)
Comment (kchugalinskiy) [reply pkg/apis/dweller/controller.go:73]: orly? did ingress authors suppose that it would be used only by them? too strict assumption, imo

File: pkg/apis/dweller/controller.go (hunk 16 above)
Comment (kchugalinskiy) [reply pkg/apis/dweller/controller.go:96]: 'just for now'

File: pkg/apis/dweller/controller.go (hunk 10 above)
Comment (kchugalinskiy) [reply pkg/apis/dweller/controller.go:122]: ok, get some function options. soon

File: pkg/apis/dweller/v1alpha1/types.go (hunk 15 above)
Comment (kchugalinskiy) [reply pkg/apis/dweller/v1alpha1/types.go:10]: got your point, I like it

File: pkg/apis/dweller/controller.go (hunk 9 above)
Comment (hypnoglow) [reply pkg/apis/dweller/controller.go:73]: > did ingress authors suppose that it would be used only by them?

I think it is not a correct example. *ingress* is a resource that is made to be used by different controllers. But our vault secret claim resource is made to be used only by our controller. If it is not, there is some flaw in design or something.
